### PR TITLE
Enable seeding actors for reproducible experiments

### PR DESCRIPTION
**Pull request by joneswong**
## What do these changes do?
Added `reproducible_seed` for `Trainer` config. Use this value to seed workers (i.e., actors) in conjunction with worker_index, so that exploration behaviors can be repeated/reproduced.

## Related issue number
#5194 

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: python/ray/rllib/evaluation/rollout_worker.py [hunk 1]
```diff
@@ -309,6 +312,9 @@ def make_env(vector_index):
                         config=tf.ConfigProto(
                             gpu_options=tf.GPUOptions(allow_growth=True)))
                 with self.tf_sess.as_default():
+                    if seed is not None:
+                        # ensure each worker has unique exploration behavior
+                        tf.set_random_seed(seed)
```
Comment (ericl) [inline python/ray/rllib/evaluation/rollout_worker.py]: Should this also call random.seed and numpy.random.seed?

Then it should be possible to write a unit test that checks two training calls with the same seed produce identical results.

File: python/ray/rllib/agents/trainer.py [hunk 2]
```diff
@@ -193,6 +193,9 @@
     # Minimum env steps to optimize for per train call. This value does
     # not affect learning, only the length of iterations.
     "timesteps_per_iteration": 0,
+    # If set to the same integer, identically configured trials will have
+    # identical random behavior. This makes experiments reproducible.
+    "reproducible_seed": None,
```
Comment (ericl) [inline python/ray/rllib/agents/trainer.py]: "experiment_seed"?

File: python/ray/rllib/evaluation/worker_set.py [hunk 3]
```diff
@@ -31,7 +32,8 @@ def __init__(self,
                  trainer_config=None,
                  num_workers=0,
                  logdir=None,
-                 _setup=True):
+                 _setup=True,
+                 seed=None):
```
Comment (ericl) [inline python/ray/rllib/evaluation/worker_set.py]: Can we make sure any underscore args always go at the end of the list, here and elsewhere?

File: python/ray/rllib/evaluation/worker_set.py [hunk 4]
```diff
@@ -53,6 +57,7 @@ def __init__(self,
         self._remote_config = trainer_config
         self._num_workers = num_workers
         self._logdir = logdir
+        self._seed = seed or int(time.time())
```
Comment (ericl) [inline python/ray/rllib/evaluation/worker_set.py]: Can the default be None? int(time.time) could have unintended correlations.

File: python/ray/rllib/evaluation/rollout_worker.py [hunk 5]
```diff
@@ -31,6 +31,7 @@
 from ray.rllib.utils.filter import get_filter
 from ray.rllib.utils.tf_run_builder import TFRunBuilder
 from ray.rllib.utils import try_import_tf
+from ray.rllib.utils.seed import seed as seed_all
```
Comment (ericl) [inline python/ray/rllib/evaluation/rollout_worker.py]: This file isn't checked in btw.

**Review by ericl — COMMENTED** [review]
Looks mostly good. Could you add a test that seeding works?

File: python/ray/rllib/agents/trainer.py [hunk 6]
```diff
@@ -429,6 +432,9 @@ def _setup(self, config):
         if self.config.get("log_level"):
             logging.getLogger("ray.rllib").setLevel(self.config["log_level"])
 
+        # Get the seed for creating workers
+        self._seed = self.config.get("seed", None)
```
Comment (richardliaw) [inline python/ray/rllib/agents/trainer.py]:
```suggestion
        self._seed = self.config.get("seed")
```

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
Merged, thanks!

**Pull request merged**